spyder/plugins/editor/widgets/codeeditor.py

```diff
@@ -1582,19 +1582,29 @@ def paste(self):
         text has 'CRLF' EOL chars
         """
         clipboard = QApplication.clipboard()
-        # This is here to prevent pasting mime data in the Editor.
-        # See issue 8566 for the details.
-        if clipboard.mimeData().hasUrls():
-            return
-        else:
-            text = to_text_string(clipboard.text())
-            if len(text.splitlines()) > 1:
-                eol_chars = self.get_line_separator()
-                text = eol_chars.join((text + eol_chars).splitlines())
-                clipboard.setText(text)
-            # Standard paste
-            TextEditBaseWidget.paste(self)
-            self.document_did_change(text)
+        text = to_text_string(clipboard.text())
+        # This is here to make copied files/folders to be pasted as text path.
+        # See issue 8566 and PR: 8576 for the details.
+        urls = []
+        if (not clipboard.mimeData().hasFormat('text/plain') and
+                clipboard.mimeData().hasUrls()):
+            urls = clipboard.mimeData().urls()
+            if len(urls) > 1:
+                text = ''.join('"' + url.toLocalFile().replace(osp.os.sep, '/')
+                               + '",' + '\n' for url in urls)
+            else:
+                text = urls[0].toLocalFile() + '\n'
+            eol_chars = self.get_line_separator()
+            text = eol_chars.join((text + eol_chars).splitlines())
+            clipboard.setText(text)
+        # Standard paste
+        TextEditBaseWidget.paste(self)
+        self.document_did_change(text)
+        # Restore files/folders into original state instead as text. See #8576.
+        if urls:

```

Comment:
That might be more [complicated](https://stackoverflow.com/questions/13762140/proper-way-to-copy-a-qmimedata-object) than it seems. As I suggested previously, all this can be avoided by not modifying the clipboard at all, and instead just pass the formatted text directly to ``TextEditBaseWidget.paste()`` (with appropriate modification to take passed text as a parameter, etc).

Reply:
> just pass the formatted text directly to `TextEditBaseWidget.paste()`

That is definitely the way to go. :+1:

Reply:
No, the purpose of this function (``CodeEditor.paste()``) is to handle pasting data, including files and paths, into a ``CodeEditor``; the fact that ``TextEditBaseWidget.paste()`` currently only pulls data from the clipboard rather than allowing it to be passed directly as a string is merely an incidental implementation detail out of convenience, and a poorly chosen one if it requires data in the clipboard to be permanently modified or lost by the simple act of pasting it.

Furthermore, I'm not sure I understand what practical objection there is to this. Adding ``text=None`` to the latter's function signature and pasting ``text`` instead of first getting the clipboard contents is a strict superset of its current functionality while still maintaining full backwards compatibility with the present invocation, so there is no real downside to adding it that I'm aware of.